### ISSUE TITLE
Customize error returned in Flask-rebar

### ISSUE THREAD

**Issue by Sytten** [plangrid/flask-rebar#79]
Today we discussed creating our own class of errors because we wanted to have uniforms errors with a machine parsable `type`. Though we could add it just fine with the additional_data field, it is a bit painful to enforce it on every error thrown. The problem is that we can't currently modify the default errors that Flask-Rebar raise in the case of an invalid schema for example. I don't have a clear answer as to how we should handle this, but I figured I could ask if someone had an idea as to how we could do this.

**Comment by barakalon** [comment]
Ah, yes. The error handling is pretty rudimentary so far.

Can you talk more about how you want to throw these errors?

Here's a quick solution that _might_ work for you:
Subclass `Rebar` and override this method: https://github.com/plangrid/flask-rebar/blob/master/flask_rebar/rebar.py#L648

```python
    def _create_json_error_response(
        self, message, http_status_code, additional_data=None, headers=None
    ):
        additional_data = additional_data or {}
        additional_data.setdefault("type", get_type_for_status_code(http_status_code))
        super()._create_json_error_response(message, http_status_code, additional_data, headers)
```

But you might deeper inspection of the original exception? I.e., `type` is specific to a `ValidationError`, not just a `BadRequest`?

**Comment by Sytten** [comment]
Yea I saw that, I think it's good practice to have machine parsable errors in general with a uniform interface. For now, what we do is using the message as our error code/type. As for the internal errors of Rebar, I just patch them:
```python
from flask_rebar import messages

messages.query_string_validation_failed = "invalid_query_string_parameters"
```
I think what I will do eventually is subclass rebar and override the `_init_error_handling` to set my own handlers.
Though one easy win would be to add a param to disable the `errors` field that is added on the bad validation. For me this has not been really helpful since it only says `missing required field` and not the field itself.

**Comment by barakalon** [comment]
> I think it's good practice to have machine parsable errors in general with a uniform interface

Totally agree. Rebar is missing this. I think this should be _in addition_ to the human readable message.

> Though one easy win would be to add a param to disable the errors field that is added on the bad validation.

I don't yet understand what this would gain?

> For me this has not been really helpful since it only says missing required field and not the field itself.

I think this was actually a limitation of marshmallow, and has been fixed with marshmallow version 3: https://github.com/plangrid/flask-rebar/blob/master/tests/test_errors.py#L281


Ok, getting back to solutions.

I think we _should_ include an "error_code" attribute in responses by default. And it should be easy to customize/add/raise error codes.

For now, maybe a root level error code is good enough? I.e., do you think its necessary to have nice, dependable error codes in the validation errors we get from marshmallow?

**Comment by Sytten** [comment]
Of right, if it was fixed then it could useful to return the `errors` field then. I think a root error would be enough. For the customization, you can always patch the field so I think if we have a sensible default that would be enough (for validation that could something like `validation_error`).

This is quite a change though, so there should be someone from plangrid that approves it. I could make a PR for the change this week or the next.

**Comment by barakalon** [comment]
> Of right, if it was fixed then it could useful to return the errors field then

I'm assuming you're rendering the errors into a form or something? 

FWIW, I think flask-rebar is used in a lot of microservice-y applications, where the HTTP status code is often enough context to programmatically respond to an error. And I think its useful to have `errors` regardless if they're "machine readable" or not, for development/debugging.

> For the customization, you can always patch the field so I think if we have a sensible default that would be enough

I generally try to avoid patching module attributes as a supported flow. Maybe we should grab the error messages from a registry that's instantiated along with `Rebar`? That may be overkill for this scenario, but it feels right to me.

**Comment by Sytten** [comment]
@RookieRick Do you have any objection?

**Comment by Sytten** [comment]
We will be using it to display an appropriate error to the user in the frontend, but I see a lot of cases where having a machine parsable error would be nice for logging/debugging inter-service communication. As for `errors`, I agree it is nice to have it.

Patching is not very nice indeed, a configurable registry would be nice. Though it is less a "priority" if we have standard machine-parsable error codes.
@RookieRick Any comment on your part?

**Comment by RookieRick** [comment]
I'm leaning pretty heavily toward the "error registry" idea, particularly if:
1. We don't change existing default behavior (duh :))
2. It opens things up for maximum customization and makes it easy for a dev to provide all of the context needed for a machine (presumably status code plus another "detail" field*) or a human (in response body).  Ideally for the case where you're going to display an error message to an end user, you should just be able to hand them the response body IMHO

* For the machine-readable "error details"  I'd be tempted to recommend this go in a response header if it's truly not for human consumption

Happy to help spitball further ideas and/or code
Thanks!!

**Comment by PratikPramanik** [comment]
For the v2.0 release, let's focus on 4XX errors
spec:
- define a schema the same format we would define a success response
- Validation error response function should return an ErrorSchema that inherits from a defined RequestSchema and adds a ErrorResponseMixin. 
- Authentication/Pre-validation error response function should return an ErrorSchema that does not inherit from RequestSchema but does add the ErrorResponseMixin
- Add an attribute to hold multiple human readable  (e.g. each text box in a form that fails validation would have a it's own error message) and machine readable fields (e.g. debug codes based on other factors that the developer defines)
- Fields have two optional parameters that inform the validation error response function: human_error and debug_error
- human_error (Optional) can be a string or a function the only returns a string.
- debug_error (Optional) is a function that returns a json parsable object.
- Validation error response function takes all of the returned stringifiable objects and mapped to field names

Sample Response
```
code: <default Marshmallow validation error code>,
message: <default Marshmallow validation error message>
errors: {
  field_name_a: { message: "this is human readable", debug: ( foo: "I can be anything" } },
  field_name_b: { message: "this is human readable as well", debug: 42 },
  field_name_c: { debug: "e3sxc39r5mn923" },
  field_name_d: { message: "Hi" }
}
```

**Comment by PratikPramanik** [comment]
This also seems like potentially v2.1 as well

**Comment by RookieRick** [comment]
Now that we're formally revisiting this one, wanted to share where my head's at and see if anybody that has been following along on this one has input on this approach:
1. Augment the existing error handling to return a unique error code alongside the human-readable message.  So for example where we currently define `body_validation_failed = "JSON body parameters are invalid."` we would instead define `body_validation_failed = ("JSON body parameters are invalid, ErrorCodes.BODY_VALIDATION_FAILED)` and surface those `ErrorCodes` values in the response we send for e.g., `BadRequest`.
2. Defer any deeper customization (i.e., field-level with error schemas that derive from request schemas) on the assumptions that a: With better errors coming from Marshmallow 3.0, we may not need them, and b: If we do need them, we can bolt that on in a 2.x release without breaking any API contracts.

If anybody (especially @Sytten @barakalon @PratikPramanik ) sees fatal flaws in that approach let me know; I'm going to spend some time today working on item 1 and hopefully we can get our 2.0 release pushed out the door soon

**Comment by RookieRick** [comment]
So, gonna tinker with this a tiny bit more before I finally make the PR, but a couple of notes..
First, as planned I have augmented the error "messages" to now contain a message and a `rebar_error_code` which results in something like the following. Note `rebar_error_code` only gets returned for errors where we have defined an error class using `HttpJsonError`. So in this example you get a 400 response with an internal code of 1 for invalid parameters, 2 for empty json body, and so on. For a 404, it behaves the same as before, with just `message`. 
I toyed with the idea of just parroting the http response code into `rebar_error_code` for cases where we didn't have an internal code, but decided that felt messy and (while a far edge case) potentially error-prone if for some reason someone defined a `rebar_error_code` of 404 that didn't mean not found.. 🤷 
![image](https://user-images.githubusercontent.com/31477753/126731366-a9e97ecd-5d25-4113-b234-28a8c3d537a2.png)

**_As I started playing with the idea of an "error handler registry" I realized we already have one..  called Flask._** 😂 
If one wants to override our default error handling, all that's needed is something like this example, in which all our built-in handlers are used except for `BadRequest`:
```    app = Flask(name)
    rebar.init_app(app)

    def my_badder_bad_request_handler(error):
        resp = flask.jsonify(
            message="I'm a liar, telling you not found for BadRequest",
            my_internal_error_code=42
        )
        resp.status_code = 404
        return resp

    app.register_error_handler(errors.BadRequest, my_badder_bad_request_handler)
```
Gonna sleep on this and if I don't dream up any reason it shouldn't be sufficient and if nobody pokes any gaping holes in this idea, will submit PR in the morning :)
